time_utils/automation/lib/python/community/time_utils.py

```diff
@@ -108,45 +115,99 @@ def to_datetime(when, log=logging.getLogger("{}.time_utils".format(LOG_PREFIX)))
     Returns:
         - DateTime specified by when
     """
+    log.debug(output)
 
     dt = None
+    dt_python = None
+    dt_java = None
+
     try:
-        # TODO import the "real" classes. Scope goes away in looping timers.
         if isinstance(when, DateTime):
+            log.debug('when is DateTime')
             dt = when
-        elif isinstance(when, (str, unicode)):
-            if is_iso8601(when):
-                dt = DateTime(when)
-            else:
-                dt = parse_duration_to_datetime(when, log)
+            dt_java = to_java_zoneddatetime(dt)
+            dt_python = to_python_datetime(dt_java)
+
         elif isinstance(when, int):
+            log.debug('when is int')
             dt = DateTime().now().plusMillis(when)
-        elif isinstance(when, (scope.DateTimeType)):
+            dt_java = ZonedDateTime.now().plus(when, ChronoUnit.MILLIS)
+            dt_python = to_python_datetime(dt_java)
+
+        elif isinstance(when, DateTimeType):
+            log.debug('when is DateTimeType')
             dt = DateTime(str(when))
-        elif isinstance(when, (scope.DecimalType, scope.PercentType,
-                            scope.QuantityType)) :
+            dt_java = to_java_zoneddatetime(dt)
+            dt_python = to_python_datetime(dt_java)
+
+        elif isinstance(when, (DecimalType, PercentType, QuantityType)):
```

Comment:
Without the scope this will likely fail when running the tests and when they are called from other libraries. For modules the scope where all the openHAB specific stuff is defined does not get imported. I see above that you've switched to importing them using the package instead of the scope. I don't remember the specific reason but Scott told me once it's better to import them using scope instead. It may be that that package will no longer exist once OH 3 comes out. And I'd like to not have to choose between OH 2 and OH 3 support at that time. But I think there was another reason to import scope and access the openHAB types that way; I just don't remember it right now.

Reply:
I thought that is what you meant by:
# TODO import the "real" classes. Scope goes away in looping timers.

reverted that change

Reply:
Sorry I didn't mean to have it so big...
While reverting that change and removing the java type libraries I ran into this issue where the scope is lost in looping timers.

Reply:
`#` turns into a Header text in these comments. In general they follow Markdown. To avoid that use code blocks:

    ```
    # TODO import the "real" classes...
    ```

Anyway, yes, I meant to remove that comment awhile back and I guess I forgot. I fought with the issue and even with the classes directly imported it wouldn't work. On the second time the timer is rescheduled the scope disappears. So I ended up creating a separate looping timer class that works differently to avoid the problem. 

So, based on the suggestions from Scott we should still use scope here.